File: apps/cargo-tracker/internal/billing/domain/discount_policy.go
package domain

// DiscountPolicy は割引率の適用ルールを表す値オブジェクト（US22）。
// 法人荷主には契約割引率を、個人荷主には割引なし（0%）を適用する。
type DiscountPolicy struct{}

// CalculateRate は荷主種別と契約割引率から適用割引率を算出する。
// 法人（CORPORATE）は契約割引率、個人（INDIVIDUAL）は 0% を返す。
func (DiscountPolicy) CalculateRate(shipperId BillingShipperId, contractRate DiscountRate) DiscountRate {
	if shipperId.IsCorporate() {
		return contractRate
	}
	return DiscountRate{rate: 0}
}

File: apps/cargo-tracker/backend/billingms/src/main/java/com/example/cargotracker/billing/domain/service/DiscountPolicy.java
package com.example.cargotracker.billing.domain.service;

import com.example.cargotracker.billing.domain.model.valueobjects.DiscountRate;
import com.example.cargotracker.billing.domain.model.valueobjects.Money;
import com.example.cargotracker.billing.domain.model.valueobjects.ShipperType;

/**
 * 法人割引（US22）。
 *
 * <p><b>種別で断ち切る。</b> 個人荷主に割引率が入っていることはありうる
 * （登録の誤り、法人から個人への変更）。率だけを見て当てると、その値がそのまま
 * 請求に効いてしまう。</p>
 *
 * <p><b>別のコマンドにしない</b>（計画の注 N10）。割引は算出の中で当てる——
 * 別コマンドにすると、割引の無い請求書が一瞬見える状態が正常系として存在する。</p>
 */
public class DiscountPolicy {

    /** 当てるべき<b>割引額</b>。当たらない種別では 0 円。 */
    public Money discountFor(ShipperType shipperType, DiscountRate rate, Money base) {
        if (!shipperType.discountable()) {
            return Money.zero();
        }
        return rate.appliedTo(base);
    }
}
